src/llm/llmnoderesources.hpp

```diff
@@ -45,62 +47,83 @@
+            return on_finalized_subword_callback(res.str());
+        }
 
-    std::optional<std::string> put(std::vector<int64_t>& tokens) {
-        for (auto token : tokens) {
-            tokenCache.push_back(token);
+        constexpr size_t delay_n_tokens = 3;
+        // In some cases adding the next token can shorten the text,
+        // e.g. when apostrophe removing regex had worked after adding new tokens.
+        // Printing several last tokens is delayed.
+        if (m_decoded_lengths.size() < delay_n_tokens) {
+            return on_finalized_subword_callback(res.str());
         }
-        std::string text = tokenizer->decode(tokenCache);
-        if (!text.empty() && '\n' == text.back() && text.size() > printLen) {
-            // The chunk is ready if the generated text ends with new line.
-            // Also, clear the cache.
-            std::string chunk = std::string{text.data() + printLen, text.size() - printLen};
-            SPDLOG_LOGGER_TRACE(llm_calculator_logger, "Generated tokens: {}", tokenCache);
-            tokenCache.clear();
-            printLen = 0;
-            return chunk;
-        } else if (!isValidUtf8(text)) {
-            return std::nullopt;
-        } else if (text.size() > printLen) {
-            // The chunk is ready if the new text in the cache contains space.
-            // The chunk is constructed from the new text, however only up to the last space character (including it)
-            // Does not clear the cache.
-            auto lastSpacePos = text.rfind(' ');
-            if (lastSpacePos == std::string::npos || lastSpacePos < printLen) {
-                return std::nullopt;
-            }
-            std::string chunk = std::string{text.data() + printLen, lastSpacePos - printLen + 1};
-            printLen = lastSpacePos + 1;
-            SPDLOG_LOGGER_TRACE(llm_calculator_logger, "Generated tokens: {}", tokenCache);
-            return chunk;
+        constexpr char replacement[] = "\xef\xbf\xbd";  // MSVC with /utf-8 fails to compile <UNK> directly with newline in string literal error.
+        if (text.size() >= 3 && text.compare(text.size() - 3, 3, replacement) == 0) {
+            m_decoded_lengths[m_decoded_lengths.size() - 1] = -1;
+            // Don't print incomplete text
+            return on_finalized_subword_callback(res.str());
         }
-        return std::nullopt;
-    }
-    std::string end() {
-        if (tokenCache.size() > 0) {
-            std::string text = tokenizer->decode(tokenCache);
-            std::string chunk = std::string{text.data() + printLen, text.size() - printLen};
-            SPDLOG_LOGGER_TRACE(llm_calculator_logger, "Generated tokens: {}", tokenCache);
-            tokenCache.clear();
-            printLen = 0;
-            return chunk;
+        int64_t print_until = m_decoded_lengths[m_decoded_lengths.size() - delay_n_tokens];
+        if (print_until != -1 && print_until > static_cast<int64_t>(m_printed_len)) {
```

Comment:
GenAI implementation does not have this static cast. We have an error when building without it.